python/xi_plugin/cache.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http:#www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import bisect
+
+
+class LineCache(object):
+    """Basic access to the core's buffer.
+
+    LineCache behaves like a list of lines. Individual lines can be accessed
+    with the lines[idx] syntax. If a line is not present in the cache,
+    it will be fetched, blocking the caller until it arrives.
+    """
+    def __init__(self, total_bytes, peer, revision, test_data=None):
+        self.total_bytes = total_bytes
+        # keep ends to make calculating offsets simple
+        self.revision = revision
+        self.peer = peer
+
+        raw_data = test_data
+        if raw_data is None:
+            raw_data = peer.get_data(0, self.revision)
+        self.raw_lines = raw_data.splitlines(True) or ['']  # handle empty buffer
+        self._recalculate_offsets()
+
+    def _recalculate_offsets(self):
+        self.offsets = [0] + [len(l) for l in self.raw_lines]
+        tally = 0
+        for idx in range(len(self.offsets)):
+            new_tally = self.offsets[idx] + tally
+            self.offsets[idx] += tally
+            tally = new_tally
+
+    def __len__(self):
+        # TODO: to report our length, in lines, we have to go get all lines, which is bad
+        while self.has_missing():
+            self.get_data(self.peer, to_offset=self.total_bytes)
+        return len(self.raw_lines)
+
+    def __getitem__(self, idx):
+        while idx >= len(self.raw_lines) and self.has_missing():
+            self.get_data(self.peer, to_offset=self.total_bytes)
+        # trim trailling newline
+        return self.raw_lines[idx].rstrip('\n')
+
+    def has_missing(self):
+        """Returns true if the cache does not have a copy of the full buffer."""
+        return self.offsets[-1] != self.total_bytes
+
+    def apply_update(self, peer, rev, start, end, new_len, edit_type, text=None):
+        # an update is bytes; can span multiple lines.
+        text = text or ""
+        assert len(text) == new_len
```

Comment:
This assert will fail if the delta is very large. The idea is that large deltas can happen, and when they do, the plugin should invalidate its contents. For a simple approach like this, I think a reasonable way to handle it is to truncate the state.